python/lsst/meas/base/dataIds.py

```diff
@@ -0,0 +1,153 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+import collections
+
+import lsst.pex.logging
+import lsst.pex.exceptions
+import lsst.afw.table
+import lsst.afw.image
+import lsst.coadd.utils
+from lsst.geom import convexHull
+
+class PerTractCcdDataIdContainer(lsst.coadd.utils.CoaddDataIdContainer):
+    """A version of lsst.pipe.base.DataIdContainer that combines raw data IDs (defined as whatever we use
+    for 'src') with a tract.
+    """
+
+    def castDataIds(self, butler):
+        """Validate data IDs and cast them to the correct type (modify idList in place).
+
+        @param butler: data butler
+        """
+        try:
+            idKeyTypeDict = butler.getKeys(datasetType="src", level=self.level)
+        except KeyError as e:
+            raise KeyError("Cannot get keys for datasetType %s at level %s: %s" % ("src", self.level, e))
+
+        idKeyTypeDict = idKeyTypeDict.copy()
+        idKeyTypeDict["tract"] = int
+
+        for dataDict in self.idList:
+            for key, strVal in dataDict.iteritems():
+                try:
+                    keyType = idKeyTypeDict[key]
+                except KeyError:
+                    validKeys = sorted(idKeyTypeDict.keys())
+                    raise KeyError("Unrecognized ID key %r; valid keys are: %s" % (key, validKeys))
+                if keyType != str:
+                    try:
+                        castVal = keyType(strVal)
+                    except Exception:
+                        raise TypeError("Cannot cast value %r to %s for ID key %r" % (strVal, keyType, key,))
+                    dataDict[key] = castVal
+
+    def _addDataRef(self, namespace, dataId, tract):
+        """Construct a dataRef based on dataId, but with an added tract key"""
+        forcedDataId = dataId.copy()
+        forcedDataId['tract'] = tract
+        dataRef = namespace.butler.dataRef(datasetType=self.datasetType, dataId=forcedDataId)
+        self.refList.append(dataRef)
+
+    def makeDataRefList(self, namespace):
+        """Make self.refList from self.idList
+        """
+        if self.datasetType is None:
+            raise RuntimeError("Must call setDatasetType first")
+        skymap = None
+        log = None
+        visitTract = {} # Set of tracts for each visit
+        visitRefs = {} # List of data references for each visit
```

Comment:
This is something that `flake8` will pick up on where it is consistent with our standard.